### ISSUE TITLE
Out-of-memory in gpuNUFFT/CUDA/inc/cuda_utils.hpp at line 40

### ISSUE THREAD

**Issue by fyrdahl** [andyschwarzl/gpuNUFFT#68]
I'm experiencing sporadic crashes with an out-of-memory error message that points towards cuda_utils.hpp line 40. Running on Nvidia Tesla K40M (12Gb GDDR5) and I'm working on 2D data, so I'm nowhere near capacity. Could this be a memory leak?

**Comment by andyschwarzl** [comment]
I'm sorry that you are experiencing crashes.

Can you give me some more details about your setup and data structure (OS, data/grid dimensions)? 

Are you using the Matlab wrapper or a standalone C++ application?

**Comment by fyrdahl** [comment]
Of course! I'm using the Matlab wrapper to grid about 36 radial spokes at the time to a 128 matrix. If by OS you mean oversampling it's 1.5 and if you mean operating system it's Ubuntu 14.04. I used a kernel size 3 and when I looked into it I noticed I was using an unusually small sector width of 3.

I ran a batch job with 40216 images reconstructed with 16 iterations of CG-SENSE, so that would be 1286912 gridding operations (forward and adjoint). In that run I experienced the "out of memory" error about 5 times, so that should give an idea of how often it occurs.

**Comment by andyschwarzl** [comment]
Thanks for sharing the information. Sector width of 3 is indeed small, but should basically work.

The batch job is quite large - are you reusing the Gridding operator instance inside Matlab? 

I've also experienced issues with huge batch jobs where GPU memory fragmentation caused these "out of memory" errors, although enough device memory had been available.

**Comment by fyrdahl** [comment]
No, 40216 unique gridding operators. I do clear them in between though. I was thinking it might be a memory leak, but you're saying about memory fragmentation makes sense.

Granted this is not a normal use case, we might be able to write it off as "shit happens" and close the issue?

**Comment by fyrdahl** [comment]
They are all unique. Binned reconstructions, the sheer amount is due to the fact that I'm batching a year worth of acquisitions.


New error that is more descriptive;

`cufft has failed at adj with err 6 
memory usage, free: 9687040 total: 12079136768
.out of memory in /home/alex/gpuNUFFT/CUDA/inc/cuda_utils.hpp at line 40`

**Comment by schrau24** [comment]
Was there ever a resolution to this? I have encountered a potentially similar out of memory issue recently when reconstructing a single 2D slice from 3D data (imageDim goes from [128 128 128] to [128 1 128]). This seems to work in most cases, however, when I repeatedly call the gpuNUFFT object it crashes and closes Matlab. I have discovered this happens when mex_gpuNUFFT_adj_atomic_f is called, but unfortunately because it quickly crashes Matlab I do not get any feedback on the what the problem is. I have tested it using the same data with different sizes and it seems to happen arbitrarily.

**Comment by andyschwarzl** [comment]
No, I didn't have the time to tackle the problem.
Can you provide me with a simple reconstruction script that (arbitrarily) reproduces the behavior?

**Comment by Mrvolcan** [comment]
I am having the same problem when reconstructing several datasets after each other. GPU memory is not fully cleared after each reconstruction. It adds up and after several datasets reaches the maximum of the GPU memory, which then leads to a crash.
Only solution so far is closing matlab after each dataset.

**Comment by fyrdahl** [comment]
I falsely believed that running `clear gpuNUFFT` would be sufficient. I later found out that I in order to release all memory, I had to clear the mex objects explicitly.
The following seemed to solve the problem for me:
`clear @gpuNUFFT/private/mex_gpuNUFFT_precomp_f;`
`clear @gpuNUFFT/private/mex_gpuNUFFT_adj_atomic_f;`

**Comment by Mrvolcan** [comment]
Thanks a lot. That solved my problem.

**Comment by andyschwarzl** [comment]
Thanks for the follow-up. To be consistent, one should also clear

`clear @gpuNUFFT/private/mex_gpuNUFFT_forw_atomic_f`

**Comment by lukasbrnst** [comment]
Hi!
Im not quiet sure if this is the right issue/thread Im reffering to.
But I have a similar problem arising:
Once I try to use '''gpuNUFFT(...)''' in the demo code, Matlab crashes very rapidly without seeing any errors when executing '''mex_gpuNUFFT_precomp_f()'''.
When observing the behaviour of my machine in the task manager, I see no rise in ram or anything else.
My machine uses 16GB ram, a Cuda 10, and Matlab 2018a.
I experience the same problem using cloud computing, so I dont think its the limitations of my machine.

Can anybody help me?

Thank you
Lukas

**Comment by curtcorum** [comment]
Working through a similar issue: 

curt@green:~/ngfn_simulation_recon/matlab$ nice matlab 
_Gtk-Message: 17:45:22.504: Failed to load module "canberra-gtk-module"_
out of memory in /home/curt/Dropbox/ngfn_simulation_recon/matlab/gpuNUFFT/CUDA/inc/cuda_utils.hpp at line 40
corrupted double-linked list
Segmentation fault (core dumped)

At this point trying to figure out if the  atomic, textures, balanced options have any effect.

    %FT_i = gpuNUFFT( kspace_i', dcf_d.^2, OS, KW, 8, imageDim, [], atomic, textures, balanced);
    FT_i = gpuNUFFT( kspace_i', dcf_d.^2, OS, KW, 8, imageDim, []);

**Comment by curtcorum** [comment]
Slightly different behavior after cleaning up arrays used to all be single. This probably does not save anything on the gpu (since being loaded as single?) but frees more memory on the cpu side.

curt@green:~/ngfn_simulation_recon/matlab$ nice matlab 
_Gtk-Message: 19:01:55.975: Failed to load module "canberra-gtk-module"_
out of memory in /home/curt/Dropbox/ngfn_simulation_recon/matlab/gpuNUFFT/CUDA/inc/cuda_utils.hpp at line 40
pure virtual method called
Segmentation fault (core dumped)


--------------------------------------------------------------------------------
          std::terminate() detected at Fri Aug 07 19:43:47 2020 -0500
--------------------------------------------------------------------------------

Configuration:
  Crash Decoding           : Disabled - No sandbox or build area path
  Crash Mode               : continue (default)
  Default Encoding         : UTF-8
  Deployed                 : false
  Desktop Environment      : GNOME-Flashback:GNOME
  GNU C Library            : 2.27 stable
  Graphics Driver          : NVIDIA Corporation Quadro RTX 8000/PCIe/SSE2 Version 4.6.0 NVIDIA 440.100
  Graphics card 1          : 0x10de ( 0x10de ) 0x1e30 Version 440.100.0.0 (0-0-0)
  Graphics card 2          : 0x10de ( 0x10de ) 0x1e30 Version 440.100.0.0 (0-0-0)
  Java Version             : Java 1.8.0_202-b08 with Oracle Corporation Java HotSpot(TM) 64-Bit Server VM mixed mode
  MATLAB Architecture      : glnxa64
  MATLAB Entitlement ID    : 4651242
  MATLAB Root              : /usr/local/MATLAB/R2019b
  MATLAB Version           : 9.7.0.1319299 (R2019b) Update 5
  OpenGL                   : hardware
  Operating System         : Ubuntu 18.04.4 LTS
  Process ID               : 105233
  Processor ID             : x86 Family 143 Model 49 Stepping 0, AuthenticAMD
  Session Key              : 0ddffd23-b786-465c-a283-d2e3cb6cfadf
  Static TLS mitigation    : Enabled: Full
  Window System            : The X.Org Foundation (12005000), display :1

Fault Count: 2


Abnormal termination:
Segmentation violation

Register State (from fault):
  RAX = 00007f9d34484d70  RBX = 00007f9c9c024170
  RCX = 00007f9d4cce3f60  RDX = 00007f9f00000001
  RSP = 00007f9fa192af08  RBP = 00007f9fa192af30
  RSI = 00007f9e21930d28  RDI = 00007f9d34484d70

   R8 = 00007f9dfc6ba974   R9 = 00007f9fac05fa90
  R10 = 0000000001d94f60  R11 = 00007f9fbdb6e9c0
  R12 = 00007f9dfc6ba800  R13 = 00007f9d73cae220
  R14 = 00007f9d4ce9b0c0  R15 = 00007f9d342fd600

  RIP = 00007f9f00000001  EFL = 0000000000010202

   CS = 0033   FS = 0000   GS = 0000

Stack Trace (from fault):
[  0] 0x00007f9f00000001                                   <unknown-module>+00000000
[  1] 0x00007f9fbdb1a0f1                    /lib/x86_64-linux-gnu/libc.so.6+00274673
[  2] 0x00007f9fa192afb0                                   <unknown-module>+00000000


Abnormal termination:
std::terminate()

Register State (captured):
  RAX = 00007f9ebf0a3000  RBX = 00007f9fbc0eed98
  RCX = 000000000000000a  RDX = 00007f9fbc0cfe00
  RSP = 00007f9ebf0a2400  RBP = 00007f9ebf0a27c0
  RSI = 00007f9fbc0972e6  RDI = 00007f9ebf0a2410

   R8 = 000000000000ffff   R9 = 29286574616e696d
  R10 = 00000000bf3fe958  R11 = 0000000000000000
  R12 = 00007f9fbc0ded08  R13 = 00007f9ebf0a2b80
  R14 = 00007f9ebf0a3070  R15 = 00007f9ebf0a3040

  RIP = 00007f9fbbfe70be  EFL = 00007f9fbc1bcbc0

   CS = 001a   FS = 0000   GS = 0000

Stack Trace (captured):
[  0] 0x00007f9fbbfde623    /usr/local/MATLAB/R2019b/bin/glnxa64/libmwfl.so+00828963
[  1] 0x00007f9fbbfde81c    /usr/local/MATLAB/R2019b/bin/glnxa64/libmwfl.so+00829468 _ZN10foundation4core4diag15stacktrace_base7captureEm+00000028
[  2] 0x00007f9fbbfe0f62    /usr/local/MATLAB/R2019b/bin/glnxa64/libmwfl.so+00839522
[  3] 0x00007f9faa759843   /usr/local/MATLAB/R2019b/bin/glnxa64/libmwmcr.so+01218627
[  4] 0x00007f9faa75c447   /usr/local/MATLAB/R2019b/bin/glnxa64/libmwmcr.so+01229895
[  5] 0x00007f9faa75c67a   /usr/local/MATLAB/R2019b/bin/glnxa64/libmwmcr.so+01230458
[  6] 0x00007f9faa75cc41   /usr/local/MATLAB/R2019b/bin/glnxa64/libmwmcr.so+01231937
[  7] 0x00007f9faa75dbc7   /usr/local/MATLAB/R2019b/bin/glnxa64/libmwmcr.so+01235911
[  8] 0x00007f9fbe16d3a6 /usr/local/MATLAB/R2019b/bin/glnxa64/../../sys/os/glnxa64/libstdc++.so.6+00582566
[  9] 0x00007f9fbe16d3f1 /usr/local/MATLAB/R2019b/bin/glnxa64/../../sys/os/glnxa64/libstdc++.so.6+00582641
[ 10] 0x00007f9fbe16debf /usr/local/MATLAB/R2019b/bin/glnxa64/../../sys/os/glnxa64/libstdc++.so.6+00585407
[ 11] 0x00007f9fbc512e26    /usr/local/MATLAB/R2019b/bin/glnxa64/libmwms.so+02113062 _ZNK10foundation7msg_svc8eventmgr9BaseEvent13getFilterTagsB5cxx11Ev+00000022
[ 12] 0x00007f9fbc5567cc    /usr/local/MATLAB/R2019b/bin/glnxa64/libmwms.so+02389964
[ 13] 0x00007f9fbc557048    /usr/local/MATLAB/R2019b/bin/glnxa64/libmwms.so+02392136
[ 14] 0x00007f9fbc5735cb    /usr/local/MATLAB/R2019b/bin/glnxa64/libmwms.so+02508235
[ 15] 0x00007f9fbc577716    /usr/local/MATLAB/R2019b/bin/glnxa64/libmwms.so+02524950
[ 16] 0x00007f9fbc577a0c    /usr/local/MATLAB/R2019b/bin/glnxa64/libmwms.so+02525708
[ 17] 0x00007f9fbc577367    /usr/local/MATLAB/R2019b/bin/glnxa64/libmwms.so+02524007
[ 18] 0x00007f9fbc55333c    /usr/local/MATLAB/R2019b/bin/glnxa64/libmwms.so+02376508
[ 19] 0x00007f9fbc551f31    /usr/local/MATLAB/R2019b/bin/glnxa64/libmwms.so+02371377
[ 20] 0x00007f9fbd4a2bdd /usr/local/MATLAB/R2019b/bin/glnxa64/libmwboost_thread.so.1.65.1+00080861
[ 21] 0x00007f9fbe4676db              /lib/x86_64-linux-gnu/libpthread.so.0+00030427
[ 22] 0x00007f9fbdbf8a3f                    /lib/x86_64-linux-gnu/libc.so.6+01186367 clone+00000063
[ 23] 0x0000000000000000                                   <unknown-module>+00000000
